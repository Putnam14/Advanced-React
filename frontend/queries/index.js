import gql from 'graphql-tag';
import { itemDetails } from './fragments';

export const CREATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation createItem($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
      ...itemDetails
    }
  }
`;

export const SIGNUP_MUTATION = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const SIGNIN_MUTATION = gql`
  mutation signup($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const REQUEST_RESET_MUTATION = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      id
    }
  }
`;

export const RESET_MUTATION = gql`
  mutation resetPassword($resetToken: String!, $password: String!, $confirmPassword: String!) {
    resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const CREATE_ORDER_MUTATION = gql`
  mutation CreateOrderMutation($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

export const ALL_ITEMS_QUERY = gql`
  # Import the Fragment
  ${itemDetails}
  query AllItemsQuery($skip: Int = 0, $first: Int = 2) {
    itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "itemsConnection") {
      aggregate {
        count
      }
    }
    items(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "items") {
      ...itemDetails
    }
  }
`;

export const SINGLE_ITEM_QUERY = gql`
  ${itemDetails}
  query FindItem($id: ID!) {
    items(where: { id: $id }) {
      user {
        id
        email
        name
      }
      ...itemDetails
    }
  }
`;

export const SINGLE_ORDER_QUERY = gql`
  query order($id: ID!) {
    order(where: { id: $id }) {
      id
      charge
      total
      createdAt
      items {
        id
        title
        price
        description
        image
        quantity
      }
    }
  }
`;

export const SEARCH_ITEMS_QUERY = gql`
  ${itemDetails}
  query SearchItems($searchTerm: String!) {
    items(
      where: {
        OR: [
          { title_contains: $searchTerm }
          { description_contains: $searchTerm }
        ]
      }
    ) {
      ...itemDetails
    }
  }
`;

export const REMOVE_ITEM_MUTATION = gql`
  mutation RemoveItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
      description
    }
  }
`;

export const UPDATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation UpdateItem(
    $id: ID!
    $title: String
    $description: String
    $price: Int
  ) {
    updateItem(
      id: $id
      description: $description
      title: $title
      price: $price
    ) {
      ...itemDetails
    }
  }
`;

export const UPDATE_USER_MUTATION = gql`
  mutation UpdateUser($name: String!) {
    updateUser(name: $name) {
      name
    }
  }
`;

export const CURRENT_USER_QUERY = gql`
  ${itemDetails}
  query userQuery {
    me {
      id
      email
      name
      orders {
        charge
        id
        total
      }
      cart {
        id
        quantity
        item {
          ...itemDetails
        }
      }
    }
  }
`;

export const USER_ORDERS_QUERY = gql`
  query orders {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      updatedAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;

export const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      item {
        id
        price
        description
        image
        title
      }
    }
  }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

export const ALL_USERS_QUERY = gql`
  query Users {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation updatePermissions($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      permissions
      name
      email
    }
  }
`;
