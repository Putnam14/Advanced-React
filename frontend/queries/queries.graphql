fragment itemDetails on Item {
  __typename
  id
  title
  price
  description
  image
  largeImage
}

mutation CREATE_ITEM_MUTATION($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
  createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
    ...itemDetails
  }
}

mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
  signup(email: $email, name: $name, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation SIGNIN_MUTATION($email: String!, $password: String!) {
  signin(email: $email, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation REQUEST_RESET_MUTATION($email: String!) {
  requestReset(email: $email) {
    id
  }
}

mutation RESET_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
  resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation CREATE_ORDER_MUTATION($token: String!) {
  createOrder(token: $token) {
    id
    charge
    total
    items {
      id
      title
    }
  }
}

query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
  itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    aggregate {
      count
    }
  }
  items(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    ...itemDetails
  }
}

query SINGLE_ITEM_QUERY($id: ID!) {
  items(where: { id: $id }) {
    user {
      id
      email
      name
    }
    ...itemDetails
  }
}

query SINGLE_ORDER_QUERY($id: ID!) {
  order(id: $id) {
    id
    charge
    total
    createdAt
    user {
      id
    }
    items {
      id
      title
      price
      description
      image
      quantity
    }
  }
}

query SEARCH_ITEMS_QUERY($searchTerm: String!) {
  items(
    where: {
      OR: [
        { title_contains: $searchTerm }
        { description_contains: $searchTerm }
      ]
    }
  ) {
    ...itemDetails
  }
}

mutation REMOVE_ITEM_MUTATION($id: ID!) {
  deleteItem(id: $id) {
    id
    title
    description
  }
}

mutation UPDATE_ITEM_MUTATION(
  $id: ID!
  $title: String
  $description: String
  $price: Int
) {
  updateItem(
    id: $id
    description: $description
    title: $title
    price: $price
  ) {
    ...itemDetails
  }
}

mutation UPDATE_USER_MUTATION($name: String!) {
  updateUser(name: $name) {
    name
  }
}

query CURRENT_USER_QUERY {
  me {
    id
    email
    name
    permissions
    orders {
      id
      charge
      total
    }
    cart {
      id
      quantity
      item {
        ...itemDetails
      }
    }
  }
}

query USER_ORDERS_QUERY {
  orders(orderBy: createdAt_DESC) {
    id
    total
    createdAt
    updatedAt
    items {
      id
      title
      price
      description
      quantity
      image
    }
  }
}

mutation ADD_TO_CART_MUTATION($id: ID!) {
  addToCart(id: $id) {
    id
    quantity
    item {
      id
      price
      description
      image
      title
    }
  }
}

mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
  removeFromCart(id: $id) {
    id
  }
}

query ALL_USERS_QUERY {
  users {
    id
    name
    email
    permissions
  }
}

mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userId: ID!) {
  updatePermissions(permissions: $permissions, userId: $userId) {
    id
    permissions
    name
    email
  }
}

query LOCAL_STATE_QUERY {
  cartOpen @client
}

mutation TOGGLE_CART_MUTATION {
  toggleCart @client
}

