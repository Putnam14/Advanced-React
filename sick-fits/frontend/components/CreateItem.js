import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  }

  // Reusable change handler for any form input
  handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  uploadFile = async e => {
    const { files } = e.target
    const data = new FormData()
    data.append('file', files[0])
    // Cloudinary upload preset
    data.append('upload_preset', 'sickfits')

    // Do the actual uploading
    const res = await fetch(
      'https://api.cloudinary.com/v1_1/sickfitsputnam/image/upload',
      {
        method: 'POST',
        body: data,
      }
    )

    // Get the files once the image has been uploaded
    const file = await res.json()
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    })
  }

  render() {
    const { title, price, description, image } = this.state
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              // Prevent page POST
              e.preventDefault()
              // Call the mutation
              const res = await createItem()
              // Send user to fresh item page
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
              </label>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  name="title"
                  id="title"
                  placeholder="Title"
                  value={title}
                  onChange={this.handleChange}
                  required
                />
                {image && <img src={image} alt="Upload preview" />}
              </label>
              <label htmlFor="price">
                Price
                <input
                  type="number"
                  name="price"
                  id="price"
                  placeholder="0.00"
                  value={price}
                  onChange={this.handleChange}
                  required
                />
              </label>
              <label htmlFor="description">
                Description
                <input
                  type="text"
                  name="description"
                  id="description"
                  placeholder="Description"
                  value={description}
                  onChange={this.handleChange}
                  required
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
