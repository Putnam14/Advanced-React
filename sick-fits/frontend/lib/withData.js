// Next-Apollo package to make sure server-side rendering works
import withApollo from 'next-with-apollo'
// You can import Apollo along with any 'link's you want, or do this
import ApolloClient from 'apollo-boost'
import { endpoint } from '../config'

function createClient({ headers }) {
  return new ApolloClient({
    // Can change this if you have different Yoga APIs for dev, prod
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          // Bring cookies along for the ride
          credentials: 'include',
        },
        headers,
      })
    },
  })
}

// Use to create client in _app.js
export default withApollo(createClient)
