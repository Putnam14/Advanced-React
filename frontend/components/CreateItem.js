import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import { CREATE_ITEM_MUTATION, ALL_ITEMS_QUERY } from '../queries/queries';
import Error from './ErrorMessage';
import Form from './styles/Form';
import PropTypes from 'prop-types';
import Router from 'next/router';
import formatMoney from '../lib/formatMoney';

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  };

  handleChange = e => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    this.setState({ loading: true });
    const files = e.currentTarget.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');

    // use the file endpoint
    const res = await fetch('https://api.cloudinary.com/v1_1/wesbos/image/upload', {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
      loading: false,
    });
  };

  // TODO: Problem here - if I refetch page 1, page 2 to infinity are still out of date. I need to refetch all the pages somehow? Or use @connection
  render() {
    return (
      <Mutation
        mutation={CREATE_ITEM_MUTATION}
        variables={this.state}
        // TODO - update all items
        // refetchQueries={[{ query: ALL_ITEMS_QUERY }]}
      >
        {(createItem, { loading, error }) => (
          <Form
            data-test
            onSubmit={async e => {
              e.preventDefault();
              const { data: { createItem: item } } = await createItem();
              Router.push({
                pathname: `/item`,
                query: { id: item.id },
              });
            }}
          >
            <h2>Sell an Item.</h2>
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label>
                Image
                <input required onChange={this.uploadFile} type="file" accept=".png, .jpg, .jpeg" />
                {this.state.image ? (
                  <img src={this.state.image} width="100" alt={this.state.title} />
                ) : null}
              </label>
              <label>
                Title
                <input
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                  type="text"
                  name="title"
                  id="title"
                  placeholder="Title"
                />
              </label>
              <label>
                Price {this.state.price && formatMoney(this.state.price)}
                <input
                  required
                  type="number"
                  id="price"
                  name="price"
                  min="0"
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>
              <textarea
                id="description"
                required
                name="description"
                value={this.state.description}
                onChange={this.handleChange}
                placeholder="The desc for this item"
              />
              <button disabled={this.state.loading} type="submit">
                Submit
              </button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
