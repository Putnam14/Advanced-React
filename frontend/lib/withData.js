import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from '../queries/queries.graphql';
// can also be a function that accepts a `headers` object (SSR only) and returns a config

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? 'http://localhost:4444' : 'http://localhost:4444',
    ssrMode: !process.browser, // Disables forceFetch on the server (so queries are only run once)
    request: async operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            const { cartOpen } = cache.read({ query: LOCAL_STATE_QUERY });
            const data = { data: { cartOpen: !cartOpen } };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo({ client: createClient });
