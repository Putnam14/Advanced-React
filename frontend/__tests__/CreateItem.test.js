import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateItem from '../components/CreateItem';
import wait from 'waait';
import { fakeItem } from '../lib/testUtils';
import { CREATE_ITEM_MUTATION } from '../queries/queries';

const dogImage = 'https://dog.com/dog.jpg';

// Mock fetch to resolve data
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [{ secure_url: dogImage }],
  }),
});

describe('<Createitem/>', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('uploads a file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const component = wrapper.find('CreateItem').instance();

    const input = wrapper.find('input[type="file"]');
    // internally this does some setState() calls
    input.simulate('change', { currentTarget: { files: ['fakedog.jpg'] } });
    await wait(0);
    expect(component.state.image).toEqual(dogImage);
    expect(component.state.largeImage).toEqual(dogImage);
  });

  it('handles state updating', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    wrapper.find('#title').simulate('change', { target: { value: 'Testing', name: 'title' } });
    wrapper.find('#price').simulate('change', {
      target: {
        value: 50000,
        name: 'price',
        type: 'number',
      },
    });
    wrapper.find('#description').simulate('change', {
      target: {
        value: 'This is a really nice item',
        name: 'description',
      },
    });
    const component = wrapper.find('CreateItem').instance();
    expect(component.state).toMatchObject({
      title: 'Testing',
      description: 'This is a really nice item',
      image: '',
      largeImage: '',
      price: 50000,
    });
  });

  it('creates an item when form is submitted', async () => {
    const item = fakeItem();
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            description: item.description,
            image: '',
            largeImage: '',
            price: item.price,
          },
        },
        result: {
          data: {
            createItem: {
              ...fakeItem(),
              __typeName: 'Item',
            },
          },
        },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    );

    wrapper.find('#title').simulate('change', { target: { value: item.title, name: 'title' } });
    wrapper.find('#price').simulate('change', {
      target: {
        value: item.price,
        name: 'price',
        type: 'number',
      },
    });
    wrapper.find('#description').simulate('change', {
      target: {
        value: item.description,
        name: 'description',
      },
    });

    Router.router = { push: jest.fn() };
    // it submits and matches the mocks
    wrapper.find('form').simulate('submit');
    // it routes to the item page
    await wait(50);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({ pathname: '/item', query: { id: '123' } });
  });
});
