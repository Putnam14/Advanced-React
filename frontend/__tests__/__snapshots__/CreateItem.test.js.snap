// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Createitem/> renders and matches snapshot 1`] = `
<form
  className="sc-bwzfXH ikCwBB"
  data-test={true}
  onSubmit={[Function]}
>
  <h2>
    Sell an Item.
  </h2>
  <DisplayError />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label>
      Image
      <input
        accept=".png, .jpg, .jpeg"
        onChange={[Function]}
        required={true}
        type="file"
      />
    </label>
    <label>
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label>
      Price 
      0
      <input
        id="price"
        min="0"
        name="price"
        onChange={[Function]}
        required={true}
        type="number"
        value={0}
      />
    </label>
    <textarea
      id="description"
      name="description"
      onChange={[Function]}
      placeholder="The desc for this item"
      required={true}
      value=""
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
